auto flushing queue client

nisanharamati opened the pull request:
`client` module
== 
Adds an `AutoFlushingQueuedClient` class to the `riemann_client.client` module with the following characteristics:
 - flush once the queue size reaches a user-defined threshold
 - flush every user-defined period (in seconds) if the queue is not empty
 - optionally disconnect the transport between flushes (default: stay connected)

`transport` module
==
Modify `BlankTransport` class to:
 - split incoming batch messages to individual messages
 - not destroy the messages collections on `connect()` and `disconnect(). Instead the collection is created at `__init__` and destroyed along with the instance when it is garbage collected.

`tests` submodule
==
Moved outside of the package to the root of the repository. 
`riemann_client.tests` removed from `setup.py` packages list.
Added `tests` to `pytest` and `flake8` sections in `tox.ini`.

README
== 
Add instructions for running tests

nisanharamati commented:
@borntyping all tests are passing, ready for review!

nisanharamati commented:
@borntyping Fixed the error handling for when the socket connection is broken. New behaviour is for the client to retry once, then clear the queue and emit an exception to the log if it fails again. It will keep retrying to connect and send though, so as soon as the remote end of the socket is back online, it will resume sending data.

borntyping commented:
Sorry, missed the original notification - I'll try and look through this today.

borntyping commented on riemann_client/transport.py:
```diff
@@ -176,20 +176,27 @@ class BlankTransport(Transport):
         """
-        self.messages.append(message)
+        for event in message.events:
+            self.messages.append(event)
```
This isn't right - `self.messages` is a list of message objects, not events.

nisanharamati replied:
I was on the fence about this one, but it makes testing for a message in a sequence of ordered batches simpler (e.g. message 153 out of 4 batches of 50 will be in `blanktransport.messages[152]`, rather than `blanktransport.messages[3].events[2]`). 
Basically I think making it look like a stream would appear in Riemann after it unpacked the batch makes it easier to test, but I'm happy to revert that if you think the transport view of batches sent is more important..

borntyping replied:
I'm fine with it storing the events instead - I don't think there's anything else Riemann would keep from a message - but the code should probably change to use `self.events` in that case so it's clear what it's doing.

borntyping commented on riemann_client/client.py:
```diff
@@ -200,3 +204,143 @@ def send_events(self, events):
+        """
+        self.send_events(self.create_event(evd) for evd in events)
+
+    def send_event(self, event):
```
This method doesn't need to be duplicated, it's already defined by `QueuedClient`.

borntyping commented on riemann_client/client.py:
```diff
@@ -200,3 +204,143 @@ def send_events(self, events):
+                                  "#2. Batch discarded.")
+                    logging.exception()
+                    self.transport.disconnect()
+                    self.clear_queue()
```
Is is a good idea to clear the queue? That could mean losing data for users. If this behaviour is keep it should be clearly documented in the class docstring.

nisanharamati replied:
There's no clearly winning choice here... you either end up with an ever-growing queue while the other end of the socket is down (so you either run out of memory, or try to send massive batches potentially affecting latencies across the entire system), or you lose messages. Eventually you'll lose messages anyway when you run out of memory and local storage capacity, though. 

How about making it tunable via an init parameter like `clear_on_fail`?

borntyping replied:
A `clear_on_fail` parameter sounds like a sensible choice.

borntyping commented on riemann_client/client.py:
```diff
@@ -6,7 +6,11 @@
+import logging
 import socket
+from threading import RLock
+from threading import Timer
```
I'm a little worried importing `threading` might break, since it's not available on on all platforms. Is there any possibility of doing this without threading?

nisanharamati replied:
Not that I know of. You can check time elapsed every time a new record is passed to the client, but you might end up with messages sitting on the queue forever if no new records come around to trigger a new flush.

nisanharamati replied:
How about making that client availability conditional on threading importing successfully?

borntyping replied:
:+1:

borntyping commented on Makefile:
```diff
@@ -0,0 +1,41 @@
+PACKAGE_NAME=riemann_client
```
Testing and build steps should use `tox` and go in `tox.ini` - this shouldn't be replicated in a Makefile.

nisanharamati commented:
- Added `clear_on_fail` flag as well as tests for it
- Cleaned up an erroneous and redundant call to `logging.exception`
- Conditional threading import logic catches `ImportError` explicitly
- Duplicated subclass methods removed
- `.messages` renamed to `.events` in `BlankTransport`.

Tests are passing.

@borntyping I think this is ready for another review?

borntyping commented:
@nisanharamati: I'm pretty happy with this, though a couple final issues:
- Testing section in the README should use `tox`.
- `AutoFlushingQueuedClient` probably shouldn't log to the root logger - using `warnings` or a child logger might be better, as I don't want to make assumptions about how the user is doing logging.

nisanharamati commented:
@borntyping 
Updated the README, added a logger instance at the head of the `client` module with a default `NullHandler`. Py 2.6 doesn't have a `NullHandler`, so I'm handling it on `ImportError`.
